WordPress/gutenberg
# Components utils: use `Object.assign` instead of `{ ...spread }` syntax

## What?


The changes in this PR refactor the code in the `utils` folder of the `@wordpress/components` package, switching to `Object.assign()` instead of using the `{ ...spread }` syntax.

## Why?


While working on #39436, I discovered the following issue, which can be replicated by running `npm run build:packages` and inspecting the following generated files:

- `packages/components/build-types/utils/config-values.d.ts`: this file actually contains a couple of circular dependency errors:

<details>
<summary>See screenshot</summary>

![Screenshot 2022-03-23 at 10 17 23](https://user-images.githubusercontent.com/1083581/159666030-090c5a6b-1e11-43c0-861b-adbbcddf5b9f.png)
</details>

- `packages/components/build-types/utils/colors-values.d.ts`: in this file, we can observe that effectively the type of the `COLORS` object doesn't include the `white` key.

## How?


In both cases, I was able to solve the issue by using `Object.assign` instead of the `{ ...spread }` syntax.

## Testing Instructions





On trunk:
- `npm run distclean && npm ci && npm run build:packages`
- Inspect `packages/components/build-types/utils/config-values.d.ts` and notice the TypeScript errors.
- Inspect `packages/components/build-types/utils/colors-values.d.ts`, and notice how the type of the `COLORS` object doesn't include the `white` key.

Apply the changes from this PR:
- `npm run distclean && npm ci && npm run build:packages`
- Inspect `packages/components/build-types/utils/config-values.d.ts`, there shouldn't be any TypeScript errors
- Inspect `packages/components/build-types/utils/colors-values.d.ts`, the type of the `COLORS` object should include the `white` key.
